torch/csrc/jit/codegen/cuda/contiguity.cpp

```diff
@@ -8,20 +9,455 @@ namespace jit {
+    consistently_ordered_ids_.emplace(merge->out());
+  }
+
+  // Don't just remove active_ids_, as if we have something like:
+  //   [i0, i1, i2, i3]
+  //   ->merge(0, 2)
+  //   ->merge(1)
+  // The latter merge looks like it's ordered correctly, if we update the active
+  // map as:
+  //   [i0, i1, i2, i3] -> [i0*i2, i1, i3]
+  // Hoever if we instead mark it as:
+  //   [i0, i1, i2, i3] -> [i0*i2, i1, nullptr, i3]
+  // Or:
+  //   [i0, i1, i2, i3] -> [nullptr, i1, i0*i2, i3]
+  // It's clear the second merge is not ordered correctly. Doesn't matter which
+  // direction we put the iter domain in, prefer putting it in outer as we often
+  // are looking for inner dimensions that are contiguous. We don't want to
+  // always do this, as it could make ordered merges look non-ordered.
+  // For exmaple: [i0, i1, i2, i3]
+  // ->merge(0)
+  // ->merge(1)
+  // ->merge(0)
+  // If it's updated as:
+  // [i0, i1, i2, i3]
+  // -> [i0*i1, nullptr, i2, i3]
+  // -> [i0*i1, nullptr, i2*i3, nullptr]
+  // Now the final merge looks non-ordered but it is. So only insert a nullptr
+  // entry if the out is not ordered.
+  active_ids_[outer_pos] = merge->out();
+
+  if (!out_ordered) {
+    active_ids_[inner_pos] = nullptr;
+  } else {
+    active_ids_.erase(active_ids_.begin() + inner_pos);
+    for (auto i = outer_pos + 1; i < inner_pos; i++) {
+      // If there's broadcast axes between outer and inner and the merge was
+      // contiguous, there may be broadcasts between outer and inner that cannot
+      // be ordered merged anywhere else so remove them.
+      active_ids_.erase(active_ids_.begin() + outer_pos + 1);
+    }
+  }
+
+  // Update the root_id entry for the output.
+  VectorOfUniqueEntries<IterDomain*> root_ids = inner_root_ids;
+  root_ids.pushBack(outer_root_ids);
+
+  id_to_root_ids_[merge->out()] = root_ids;
+
+  // Need to check this after updating active_ids_ and id_to_root_ids_
+  if (checkExclusivelyConsumesRoots(merge->out())) {
+    exclusively_consumes_roots_.emplace(merge->out());
+  }
+}
+
+void OrderedIdInformation::handle(Split* split) {
+  // Find the input in the active_ids_ vector
+  const auto in_it =
+      std::find(active_ids_.begin(), active_ids_.end(), split->in());
+
+  if (in_it == active_ids_.end()) {
+    return;
+  }
+
+  auto in_pos = std::distance(active_ids_.begin(), in_it);
+
+  // Find the input in the ordered transforms map
+  const auto in_ordered_it = consistently_ordered_ids_.find(split->in());
+
+  bool in_ordered = in_ordered_it != consistently_ordered_ids_.end();
+
+  // Get root ids of the input
+  const auto in_root_ids_it = id_to_root_ids_.find(split->in());
+
+  TORCH_INTERNAL_ASSERT(
+      in_root_ids_it != id_to_root_ids_.end(),
+      "Error replaying transforms in contiguous ID checker.");
+
+  VectorOfUniqueEntries<IterDomain*> in_root_ids = in_root_ids_it->second;
+
+  // Need to check this before we update the active_ids_
+
+  if (checkExclusivelyConsumesRoots(split->in())) {
```

Comment:
Seems odd to do this for inputs. If inputs are outputs of a merge, this would be redundant.

Reply:
Yeah, I think you're right. Since we initialize this for roots to be true, and check on merge outputs, there shouldn't be any other cases that can be true, so checking here should be purely redundant.